### PR TITLE
`Adaptive learning`: Add import of standardized competencies from .json files

### PR DESCRIPTION
**Pull request by rstief**
# Only test on ts3 (or another server where you have admin permissions)

#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
Adds the functional requirement 2.1 described in #8001 

### Description
Adds the option to import standardized competencies from .json files into an Artemis instance.

### Steps for Testing
Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Go to Server Administration -> Feature Toggles and activate Standardized Competencies
3. Go to Server Administration -> Standardized Competencies -> Import
4. Use the following files to test the import (or create your own if you dare 😉 )
  Malformed json syntax: [import_malformed.json](https://github.com/ls1intum/Artemis/files/14972404/import_malformed.json)
  Invalid objects: [invalid.json](https://github.com/ls1intum/Artemis/files/14972403/invalid.json)
  Correct: [import_correct.json](https://github.com/ls1intum/Artemis/files/14972405/import_correct.json)
6. Verify that the new competencies got imported (in the Standardized Competency Management view)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [admin-standardized-competency.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/standardized-competencies/admin-standardized-competency.service.ts.html) | 100% | ✅ |
| [admin-import-standardized-competencies.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.ts.html) | 97.18% | ✅  |
| [knowledge-area-detail.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/standardized-competencies/knowledge-area-detail.component.ts.html) | 98.41% | ✅  |
| [knowledge-area-tree.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/standardized-competencies/knowledge-area-tree/knowledge-area-tree.component.ts.html) | 100% | ✅ |
| [standardized-competency-detail.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/standardized-competencies/standardized-competency-detail.component.ts.html) | 100% | ✅ |
| [standardized-competency-management.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/standardized-competencies/standardized-competency-management.component.ts.html) | 88.5% | ✅ |
| [standardized-competency.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/competency/standardized-competency.model.ts.html) | 100% | ✅ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 89.69% | ✅  |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [KnowledgeArea.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/KnowledgeArea.html) | 100% | ✅ |
| [StandardizedCompetency.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/StandardizedCompetency.html) | 89% | ✅  |
| [KnowledgeAreaRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.competency/KnowledgeAreaRepository.html) | 100% | ✅ |
| [KnowledgeAreaService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.competency/KnowledgeAreaService.html) | 100% | ✅ |
| [StandardizedCompetencyService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.competency/StandardizedCompetencyService.html) | 97% | ✅  |
| [AdminStandardizedCompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminStandardizedCompetencyResource.html) | 100% | ✅ |
| [StandardizedCompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.competency/StandardizedCompetencyResource.html) | 100% | ✅ |
| [KnowledgeAreaRequestDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.standardizedCompetency/KnowledgeAreaRequestDTO.html) | 100% | ✅ |
| [KnowledgeAreaResultDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.standardizedCompetency/KnowledgeAreaResultDTO.html) | 100% | ✅ |
| [KnowledgeAreasForImportDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.standardizedCompetency/KnowledgeAreasForImportDTO.html) | 100% | ✅ |
| [StandardizedCompetencyRequestDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6402/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.standardizedCompetency/StandardizedCompetencyRequestDTO.html) | 100% | ✅ |

### Screenshots
Display before selecting a file
image
Preview after selecting a file
image

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive module for importing standardized competencies, including knowledge areas and sources, with a detailed user interface.
	- Added new components for managing and displaying a tree structure of knowledge areas in the admin section.
	- Enhanced user interface for importing competencies, providing options for file management, data preview, and import operations.
  
- **Enhancements**
	- Updated the standardized competencies management interface to include new import functionalities and improved layout and styling.
  
- **Documentation and Localization**
	- Expanded user guidance for competency import processes in both English and German, including detailed error and success message handling.

- **Tests**
	- Implemented a robust suite of tests for new components and services, ensuring functionality and error handling are up to standard.

### CODE REVIEW

**Review by mateusmm01 — DISMISSED** [review]
Worked as expected in testing session

**Review by laurenzfb — DISMISSED** [review]
demonstrated in testing session, looks good

**Review by florian-glombik — APPROVED** [review]
Tested in testing session, validation worked for invalid file extensions, invalid format and a valid file.

We also discussed if it makes sense to further enhance the error message in the case of an invalid format (displaying the line in which the parsing error occurs) but decided that this should not be needed for now, as the file will be exported and imported for the usual use case anyways and further dependencies would be necessary.

File: src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.html [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        <div class="card-body" [ngbCollapse]="isCollapsed">
+            <p jhiTranslate="artemisApp.standardizedCompetency.manage.import.format.text1"></p>
+            <p jhiTranslate="artemisApp.standardizedCompetency.manage.import.format.text2"></p>
+            <p jhiTranslate="artemisApp.standardizedCompetency.manage.import.format.text3"></p>
+            <div [innerHTML]="importExample | htmlForMarkdown"></div>
+        </div>
+    </div>
+    <input class="mt-3" type="file" accept="application/json" (change)="onFileChange($event)" />
+    <h4 class="mt-3" jhiTranslate="artemisApp.standardizedCompetency.manage.import.preview.title"></h4>
+    @if (count && count?.knowledgeAreas) {
```
Comment (pzdr7) [inline src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.html:19]: Maybe you could rename this field to `importCount` - just something I stumbled over when reviewing

File: src/test/javascript/spec/component/standardized-competencies/admin-import-standardized-competencies.spec.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    });
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    it('should initialize', () => {
+        componentFixture.detectChanges();
+        expect(component).toBeDefined();
+    });
```
Comment (pzdr7) [inline src/test/javascript/spec/component/standardized-competencies/admin-import-standardized-competencies.spec.ts:42]: This test can be removed

File: src/test/javascript/spec/component/standardized-competencies/admin-import-standardized-competencies.spec.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            result: JSON.stringify(result),
+        } as FileReader;
+        const expectedCount = { knowledgeAreas: 4, competencies: 5 };
+
+        component['setImportDataAndCount']();
+
+        expect(component['count']).toEqual(expectedCount);
+    });
+
+    it('should import competencies', () => {
```
Comment (pzdr7) [inline src/test/javascript/spec/component/standardized-competencies/admin-import-standardized-competencies.spec.ts:130]: This test only seems to check whether we navigate. Maybe update the description

**Review by pzdr7 — COMMENTED** [review]
Code looks good. Just a few small remarks :+1:

**Review by Jan-Thurner — APPROVED** [review]
lgtm

**Review by Jan-Thurner — APPROVED** [review]

**Review by laurenzfb — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**